types/angular/index.d.ts

```diff
@@ -1029,8 +1029,7 @@ declare namespace angular {
          *
          * @param promises A hash of promises.
          */
-        all(promises: { [id: string]: IPromise<any>; }): IPromise<{ [id: string]: any; }>;
-        all<T extends {}>(promises: { [id: string]: IPromise<any>; }): IPromise<T>;
+        all<T extends {}>(promises: { [K in keyof T]: IPromise<T[K]>; }): IPromise<T>;
```

Comment:
Good points, thanks for having a look. I have added a few tests. 

I have also fixed the signature so the function also takes non-promise properties.